File: 0x00基本算法/0x02递推与递归/AcWing93.cpp
#include <iostream>
#include <vector>
using namespace std;
vector<int> chosen;
int stk[100010], top = 0, address = 0, n, m;

void call(int x, int ret_addr) {
  int old_top = top;
  stk[++top] = x;
  stk[++top] = ret_addr;
  stk[++top] = old_top;
}

int ret() {
  int ret_addr = stk[top-1]; // 恢复返回地址
  top = stk[top]; // 恢复旧的top位置
  return ret_addr;
}

int main() {
  cin >> n >> m;
  call(1, 0);
  while (top) {
    int x = stk[top-2]; // call中放入栈的参数x
    switch (address) {
    case 0: //对应原递归函数入口
      if (chosen.size()>m || chosen.size()+(n-x+1)<m) {
        address = ret();
        continue;
      }
      if (x == n+1) {
        for (int i=0; i<chosen.size(); i++)
          cout << chosen[i] << " ";
        cout << endl;
        address = ret();
        continue;
      }
      chosen.push_back(x); // 准备选当前的x放入chosen
      call(x+1, 1); // 相当于calc(x+1)，返回后从case 1继续
      address = 0;
      continue; // 开启选x的新递归
    case 1: // 对应原递归函数在选完x之后
      chosen.pop_back();
      call(x+1, 2); // 相当于calc(x+1)，返回后从case 2继续
      address = 0;
      continue; // 开启不选x的新递归
    case 2: //对应原递归函数在不选择x之后
      address = ret();
    }
  }
  return 0;
}

File: 0x00 基本算法/0x02 递推与递归/AcWing93.py
stk, top, address = [0]*100010, 0, 0
chosen = []
def calc(x, ret_addr):
    global top, stk
    old_top = top
    top += 1; stk[top] = x
    top += 1; stk[top] = ret_addr
    top += 1; stk[top] = old_top

def ret():
    global top, stk
    ret_addr = stk[top-1]
    top = stk[top]
    return ret_addr
    
def main():
    global stk, top, address
    n, m = [int(x) for x in input().split()]
    calc(1, 0)
    while top:
        x = stk[top-2]
        if address == 0:
            if len(chosen) > m or len(chosen)+(n-x+1) < m:
                address = ret()
                continue
            if x == n+1:
                for i in range(len(chosen)):
                    print("%d " %chosen[i], end='')
                print()
                address = ret()
                continue
            chosen.append(x)
            calc(x+1, 1)
            address = 0
            continue
        elif address == 1:
            chosen.pop()
            calc(x+1, 2)
            address = 0
            continue
        else:
            address = ret()
            
main()